cpp/kiss_icp/core/VoxelHashMap.cpp

```diff
@@ -43,42 +43,32 @@ std::vector<Voxel> GetAdjacentVoxels(const Voxel &voxel, int adjacent_voxels = 1
     }
     return voxel_neighborhood;
 }
-std::vector<Eigen::Vector3d> GetPoints(const std::vector<Voxel> &query_voxels,
-                                       const kiss_icp::VoxelHashMap &voxel_map) {
-    std::vector<Eigen::Vector3d> points;
-    points.reserve(query_voxels.size() * static_cast<size_t>(voxel_map.max_points_per_voxel_));
-    std::for_each(query_voxels.cbegin(), query_voxels.cend(), [&](const auto &query) {
-        auto search = voxel_map.map_.find(query);
-        if (search != voxel_map.map_.end()) {
-            const auto &voxel_points = search.value();
-            points.insert(points.end(), voxel_points.cbegin(), voxel_points.cend());
-        }
-    });
-    points.shrink_to_fit();
-    return points;
-}
-
 }  // namespace
 
 namespace kiss_icp {
 
 std::tuple<Eigen::Vector3d, double> VoxelHashMap::GetClosestNeighbor(
-    const Eigen::Vector3d &point) const {
+    const Eigen::Vector3d &query) const {
     // Convert the point to voxel coordinates
-    const auto &voxel = PointToVoxel(point, voxel_size_);
+    const auto &voxel = PointToVoxel(query, voxel_size_);
     // Get nearby voxels on the map
     const auto &query_voxels = GetAdjacentVoxels(voxel);
-    // Extract the points contained within the neighborhood voxels
-    const auto &neighbors = GetPoints(query_voxels, *this);
-
     // Find the nearest neighbor
     Eigen::Vector3d closest_neighbor = Eigen::Vector3d::Zero();
     double closest_distance = std::numeric_limits<double>::max();
-    std::for_each(neighbors.cbegin(), neighbors.cend(), [&](const auto &neighbor) {
-        double distance = (neighbor - point).norm();
-        if (distance < closest_distance) {
-            closest_neighbor = neighbor;
-            closest_distance = distance;
+    auto closer = [&query](const auto &lhs, const auto &rhs) {
```

Comment:
Why not throw this lambda in the `min_element`  search? It isn't very clear for me to `closer`, and then I had to look where this was used. 

On the other hand, is not very clear what is going on here, too much `auto`, `lhs`, Later on `const auto &neighbor = *std::min_element(points.cbegin(), points.cend(), closer);`.... If we want to go this road I will use more explicit types or try to rewrite this in a clear manner.

Reply:
I would also vote for moving it below, then we can also avoid re-computing the distance to the closest neighbor

Reply:
Done, tell me if this is better